### DISCUSSION TITLE
Problem: Outsized EBSD map

### DISCUSSION THREAD

**Discussion by caoyi0701** [mtex-toolbox/mtex#589] [Ask Anything] [answered]
Hi all,

I am attempting to calcuate and plot the grains of an EBSD map using the function below.

`[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'threshold',[1*degree, 10*degree],'boundary','tight');
ebsd(grains(grains.grainSize<=2)) = [];
F = KuwaharaFilter;
F.numNeighbours = 5;
ebsd = smooth(ebsd('indexed'),F);
[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'threshold',[1*degree, 10*degree],'boundary','tight');
plot(grains)`

However, the figure it yielded turns out to be strange as shown in the attached figure.
![Phase map_example](https://user-images.githubusercontent.com/75005172/100187706-a3005180-2f23-11eb-81b1-b14d41d1e876.png)
The frame of the figure is much larger than the original EBSD map.
Could anyone help fix this problem？

Best regards,
Yi Cao

**Comment by ralfHielscher** [comment]
Hi Yi

 1. Could you please attache your data?
 2. Could you identify it which step during your calculations the error occurs? 

Ralf.

**Reply by caoyi0701** [reply]
Hi Ralf,

Thank you for the reply.
Please see attachment for my data.
[example EBSD data.zip](https://github.com/mtex-toolbox/mtex/files/5594805/example.EBSD.data.zip)

I selected a small region to plot the figure as shown previously, using the codes below.

`%% Rotate EBSD data 
% roate the whole data set about the z-axis by 90*degree
ebsd = rotate(ebsd,-90*degree);
plot(ebsd)

%% Remove inaccurate orientations
figure(1)
ebsd = ebsd(ebsd.mad<1);
plot(ebsd,ebsd.mad)
mtexColorbar
%
figure(2)
plot(ebsd)

%% Select regions
% Select Bt-poor layer (BPL)
figure(3)
region1 = [-933 -17726; -2164 -17805; -3401 -17943; -4633 -17894; -5935, -18758; -7206 -19185; -7305 -19324;...
    -8397 -19393; -9468 -19751; -10404 -20506; -10812 -20933; -11414 -20981; -11656 -21251; -12103 -21271; -12327 -20981;...
    -12803 -20704; -13121 -20465; -13518 -20346; -14409 -20482; -15260 -20777; -17250 -21126;-17250 -26142;...
    -14550 -25874; -12508 -25705; -10990 -25334; -9201 -24963; -7582 -24355; -5894 -23950; -3712 -23324;-1012 -23006];
ebsd = ebsd(inpolygon(ebsd,region1))
plot(ebsd)`

When plotting the EBSD map, i.e., plot(ebsd), the figure is okay. But the problem occurs when plotting the grains, using plot(grains).

Best regards,
Yi Cao

**Reply by ralfHielscher** [reply]
Hi Cia,

unfortunately I was not able to reproduce  your error. The are two problems with you code:

 1. no grains are computed that could be plotted
 2. after restricting to the region the ebsd variable is empty

Ralf.

**Reply by caoyi0701** [reply]
Hi Ralf,

I attached the complete code of my script here, which I tested. I hope it can work.

`%% Import Script for EBSD Data
%
% This script was automatically created by the import wizard. You should
% run the whoole script or parts of it in order to import your data. There
% is no problem in making any changes to this script.

%% Specify Crystal and Specimen Symmetries

% crystal symmetry
CS = {... 
  'notIndexed',...
  crystalSymmetry('321', [4.9 4.9 5.5], 'X||a*', 'Y||b', 'Z||c*', 'mineral', 'Quartz', 'color', [0 1 0]),...
  crystalSymmetry('-1', [8.2 13 14], [93.51,115.91,90.66]*degree, 'X||a*', 'Z||c', 'mineral', 'Bytownite', 'color', [0 0 1]),...
  crystalSymmetry('4/mmm', [4.6 4.6 3], 'mineral', 'Rutile', 'color', [1 0 0]),...
  crystalSymmetry('12/m1', [5.3 9.3 10], [90,100.23,90]*degree, 'X||a*', 'Y||b*', 'Z||c', 'mineral', 'Biotite', 'color', [0.85 0.65 0.13]),...
  crystalSymmetry('m-3m', [12 12 12], 'mineral', 'Almandine', 'color', [1 0.75 0.8]),...
  crystalSymmetry('-3', [5.1 5.1 14], 'X||a*', 'Y||b', 'Z||c*', 'mineral', 'Ilmenite', 'color', [0 0.39 0]),...
  crystalSymmetry('m-3m', [8.4 8.4 8.4], 'mineral', 'Magnetite', 'color', [0.55 0 0.83])};

% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','outOfPlane');

%% Specify File Names

% path to files
pname = 'C:\Users\DELL\Desktop\example.EBSD.data\example EBSD data';

% which files to be imported
fname = [pname '\example EBSD data.cpr'];

%% Import the Data

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertSpatial2EulerReferenceFrame');

%% Rotate EBSD data 
% roate the whole data set about the z-axis by 80*degree
ebsd = rotate(ebsd,-90*degree);
plot(ebsd)

%% Remove inaccurate orientations
figure(1)
ebsd = ebsd(ebsd.mad<1);
plot(ebsd,ebsd.mad)
mtexColorbar
%
figure(2)
plot(ebsd)

%% Select regions
% Select Bt-poor layer (BPL)
figure(3)
region1 = [-933 -17726; -2164 -17805; -3401 -17943; -4633 -17894; -5935, -18758; -7206 -19185; -7305 -19324;...
    -8397 -19393; -9468 -19751; -10404 -20506; -10812 -20933; -11414 -20981; -11656 -21251; -12103 -21271; -12327 -20981;...
    -12803 -20704; -13121 -20465; -13518 -20346; -14409 -20482; -15260 -20777; -17250 -21126;-17250 -26142;...
    -14550 -25874; -12508 -25705; -10990 -25334; -9201 -24963; -7582 -24355; -5894 -23950; -3712 -23324;-1012 -23006];
ebsd = ebsd(inpolygon(ebsd,region1))
plot(ebsd)

%% Calculate grains 
[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'threshold',[1*degree, 10*degree],'boundary','tight');
ebsd(grains(grains.grainSize<=2)) = [];
%[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'threshold',[1*degree, 10*degree],'boundary','tight');
F = KuwaharaFilter;
F.numNeighbours = 5;
ebsd = smooth(ebsd('indexed'),F);
[grains,ebsd('indexed').grainId,ebsd.mis2mean] = calcGrains(ebsd('indexed'),'threshold',[1*degree, 10*degree],'boundary','tight');
% Smooth grains
grains_smoothed = smooth(grains,4); % "grains_smoothed_temp" is raw constructed grains; it is used for merging twin boundaries
plot(grains_smoothed)
`

All the best,
Yi

**Comment by kilir** [comment]
If I remember correctly, some older version of mtex had a bug after smoothing where under certain conditions some coordinates were placed at funny conditions - but that's maybe +1/2 year ago. Do you use a recent version? 
Rüdiger

**Reply by caoyi0701** [reply]
Hi Rüdiger,

Thank you for the reply. I am using the latest version of MTEX, ver. 5.5.0.

Yi

**Comment by kilir** [comment]
Hi Yi,
your region seems to be a bit ill defined, hence one cannot reproduce your issue right way. I think in your no fully correct mixture of `ebsd` and ebsd('indexed')` some things may go wrong: Using the following works for me:

```
CS = {... 
  'notIndexed',...
  crystalSymmetry('321', [4.9 4.9 5.5], 'X||a*', 'Y||b', 'Z||c*', 'mineral', 'Quartz-new', 'color', [0.53 0.81 0.98]),...
  crystalSymmetry('-1', [8.2 13 14], [93.51,115.91,90.66]*degree, 'X||a*', 'Z||c', 'mineral', 'Bytownite', 'color', [0.56 0.74 0.56]),...
  crystalSymmetry('4/mmm', [4.6 4.6 3], 'mineral', 'Rutile', 'color', [0.85 0.65 0.13]),...
  crystalSymmetry('12/m1', [5.3 9.3 10], [90,100.23,90]*degree, 'X||a*', 'Y||b*', 'Z||c', 'mineral', 'Biotite', 'color', [0.94 0.5 0.5]),...
  crystalSymmetry('m-3m', [12 12 12], 'mineral', 'Almandine', 'color', [0 0 0.55]),...
  crystalSymmetry('-3', [5.1 5.1 14], 'X||a*', 'Y||b', 'Z||c*', 'mineral', 'Ilmenite', 'color', [0 0.39 0]),...
  crystalSymmetry('m-3m', [8.4 8.4 8.4], 'mineral', 'Magnetite', 'color', [0.55 0 0])};

% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','intoPlane');
fname = ['example EBSD data.crc'];

ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertEuler2SpatialReferenceFrame');
ebsd = rotate(ebsd,-90*degree);

region1 = [-933 -17726; -2164 -17805; -3401 -17943; -4633 -17894; -5935, -18758; -7206 -19185; -7305 -19324;...
-8397 -19393; -9468 -19751; -10404 -20506; -10812 -20933; -11414 -20981; -11656 -21251; -12103 -21271; -12327 -20981;...
-12803 -20704; -13121 -20465; -13518 -20346; -14409 -20482; -15260 -20777; -17250 -21126;-17250 -26142;...
-14550 -25874; -12508 -25705; -10990 -25334; -9201 -24963; -7582 -24355; -5894 -23950; -3712 -23324;-1012 -23006];

region1 = [-region1(:,1) region1(:,2)];

plot(ebsd)
hold on
plot(region1(:,1),region1(:,2))
hold off

ebsd = ebsd(inpolygon(ebsd,region1))

[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'threshold',[1*degree, 10*degree],'boundary','tight');
ebsd(grains(grains.grainSize<=2)) = [];
F = KuwaharaFilter;
F.numNeighbours = 5;
ebsd = smooth(ebsd,F);
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'threshold',[1*degree, 10*degree],'boundary','tight');
grains_smoothed = smooth(grains,4);
plot(grains_smoothed)
```
Cheers,
Rüdiger

**Reply by caoyi0701** [reply]
Hi Rüdiger, 

Thanks a lot for your reply.

I tested your codes, it works also well for ebsd('indexed'). I feel that my error occurred probably because of the choice of reference frame, since I used 'convertSpatial2EulerReferenceFrame'. 

PS., This is a bit of strange, because I also used such reference frame for my other EBSD data, but there is no such error.

All the best,
Yi

**Reply by kilir** [reply]
Hi Yi,
no, that's unrelated. As far as I can see it is a bug in `calcGrains` due to mixing of non-existent and nonIndexed points in combiantion with the tight boundary option.

Here's where it occurs:
```
CS = {... 
  'notIndexed',...
  crystalSymmetry('321', [4.9 4.9 5.5], 'X||a*', 'Y||b', 'Z||c*', 'mineral', 'Quartz-new', 'color', [0.53 0.81 0.98]),...
  crystalSymmetry('-1', [8.2 13 14], [93.51,115.91,90.66]*degree, 'X||a*', 'Z||c', 'mineral', 'Bytownite', 'color', [0.56 0.74 0.56]),...
  crystalSymmetry('4/mmm', [4.6 4.6 3], 'mineral', 'Rutile', 'color', [0.85 0.65 0.13]),...
  crystalSymmetry('12/m1', [5.3 9.3 10], [90,100.23,90]*degree, 'X||a*', 'Y||b*', 'Z||c', 'mineral', 'Biotite', 'color', [0.94 0.5 0.5]),...
  crystalSymmetry('m-3m', [12 12 12], 'mineral', 'Almandine', 'color', [0 0 0.55]),...
  crystalSymmetry('-3', [5.1 5.1 14], 'X||a*', 'Y||b', 'Z||c*', 'mineral', 'Ilmenite', 'color', [0 0.39 0]),...
  crystalSymmetry('m-3m', [8.4 8.4 8.4], 'mineral', 'Magnetite', 'color', [0.55 0 0])};

% plotting convention
setMTEXpref('xAxisDirection','east');
setMTEXpref('zAxisDirection','intoPlane');
fname = ['example EBSD data.crc'];

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','crc',...
  'convertSpatial2EulerReferenceFrame');

ebsd = rotate(ebsd,-90*degree);

region1 = [-933 -17726; -2164 -17805; -3401 -17943; -4633 -17894; -5935, -18758; -7206 -19185; -7305 -19324;...
-8397 -19393; -9468 -19751; -10404 -20506; -10812 -20933; -11414 -20981; -11656 -21251; -12103 -21271; -12327 -20981;...
-12803 -20704; -13121 -20465; -13518 -20346; -14409 -20482; -15260 -20777; -17250 -21126;-17250 -26142;...
-14550 -25874; -12508 -25705; -10990 -25334; -9201 -24963; -7582 -24355; -5894 -23950; -3712 -23324;-1012 -23006];

ebsd = ebsd(inpolygon(ebsd,region1));

plot(ebsd)
grains = calcGrains(ebsd('indexed'),'threshold',10*degree,'boundary','tight');
hold on
plot(grains.boundary)
hold off
```

![no](https://user-images.githubusercontent.com/11697557/100629656-d5e28500-3329-11eb-98d8-755814a728a6.jpg)

and here's one way to work around the problem:
```
ebsd = ebsd.gridify;
ebsd(isnan(ebsd.phaseId)).phaseId=1;

plot(ebsd)
grains = calcGrains(ebsd('indexed'),'threshold',10*degree,'boundary','tight');
hold on
plot(grains.boundary)
hold off
```

![yes](https://user-images.githubusercontent.com/11697557/100629687-df6bed00-3329-11eb-9129-c99a6774393d.jpg)


Hope that helps for now.

Cheers,
Rüdiger

**Reply by caoyi0701** [reply]
Hi Rüdiger,

Thank you very much for your patient help.

I used the alternative approach you provided and it works well for now. But I also found that after removing the high MAD data points using `ebsd = ebsd(ebsd.mad<1);`, the result still has an error.

For a comparison, I attached the results of both with and without high MAD data points removal.

Phase map with all MAD points:
![Phase map with all MAD points](https://user-images.githubusercontent.com/75005172/100690772-3a512300-33c2-11eb-8689-d37f67527854.jpg)

Phase map with only low MAD points:
![Phase map with low MAD points](https://user-images.githubusercontent.com/75005172/100690791-450bb800-33c2-11eb-9208-7074a2d6e7fe.jpg)

Nevertheless, this result is good enough for my purpose; and I am looking forward to an update on this `calcGrains` function in the future.
Thank you again for your kind help.

Best wishes,
Yi

**Reply by kilir** [reply]
Hi,
Did you try to fix your nonIndexed points after removing those pixels?
Cherrs, 
Rüdiger

**Reply by caoyi0701** [reply]
Hi Rüdiger,

Yes, I did. I run the following codes `ebsd = ebsd.gridify; ebsd(isnan(ebsd.phaseId)).phaseId = 1;` after removing the high MAD pixels using `ebsd = ebsd(ebsd.mad<1);`. The result turned out to be which I attached in my previous reply.

Best wishes,
Yi

**Reply by kilir** [reply]
Same issue as you create non existent pixels, `ebsd(ebsd.mad<1).phaseId=1;` works for me.

Cheers,
Rüdiger

**Reply by caoyi0701** [reply]
Hi Rüdiger,

The code `ebsd(ebsd.mad<1).phaseId=1` also works for me. But does it set the low mad pixels (mad<1) as notIndexed or non-existent pixels? When I try the code `ebsd(ebsd.mad>1).phaseId=1`, the result is the same as before.

Best wishes,
Yi

**Comment by ralfHielscher** [comment, marked as answer]
Hi Yi,

I did some testing and assume that this problem is due to a bug in QHull, which is a library for Voronoi tessellation used by MTEX.
For me everything was fine when I used the option `noQHull` which causes MTEX to use a different library for this task, i.e., try
```matlab
grains = calcGrains(ebsd('indexed'),'threshold',10*degree,'boundary','tight','noQHull');
```
Ralf.

**Reply by caoyi0701** [reply]
Hi Ralf,

Thanks a lot for your solution. The code also works fine for me.

Best wishes,
Yi
